### PR TITLE
HAAR-686 Limit direct access to URLs by checking roles in token

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

**Comment by karenmillermoj** [comment]
I still need to update the cypress tests to add a test to mimic the direct url access returning an authError

File: backend/index.js [hunk 1]
```diff
@@ -18,7 +19,7 @@ require('express-async-errors')
 const apis = require('./apis')
 const ensureHttps = require('./middleware/ensureHttps')
 const healthFactory = require('./services/healthCheck')
-const setupAuth = require('./setupAuth')
+const setupAuth = require('./middleware/setUpAuth')
```
Comment (karenmillermoj) [inline backend/index.js:22]: Just moved to a new place to match template project

File: backend/middleware/authorisationMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  return (req, res, next) => {
+    if (res.locals?.access_token) {
+      const { authorities: roles = [] } = jwtDecode(res.locals.access_token) as { authorities?: string[] }
+
+      if (authorisedRoles.length && !roles.some((role) => authorisedRoles.includes(role))) {
+        logger.error('User is not authorised to access this')
+        return res.redirect('/authError')
+      }
+
+      if (req.originalUrl.endsWith('/select-caseloads') && !roles.includes('ROLE_MAINTAIN_ACCESS_ROLES_ADMIN')) {
```
Comment (karenmillermoj) [inline backend/middleware/authorisationMiddleware.ts:15]: This is the line stopping direct access to the url

File: backend/middleware/authorisationMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import jwtDecode from 'jwt-decode'
+import type { RequestHandler } from 'express'
+import logger from '../log'
+
+export default function authorisationMiddleware(authorisedRoles: string[] = []): RequestHandler {
+  return (req, res, next) => {
+    if (res.locals?.access_token) {
+      const { authorities: roles = [] } = jwtDecode(res.locals.access_token) as { authorities?: string[] }
+
+      if (authorisedRoles.length && !roles.some((role) => authorisedRoles.includes(role))) {
```
Comment (karenmillermoj) [inline backend/middleware/authorisationMiddleware.ts:10]: no authorisedRoles are passed in at the moment - as this file grows we can possibly specify the only ones allowed

**Review by Mjwillis — APPROVED** [review]
LGTM

**Pull request merged**